mypy/nodes.py

```diff
@@ -383,7 +383,18 @@ def __str__(self) -> str:
 ]  # type: Final
 
 
-class FuncBase(Node):
+class FuncBaseMeta(type):
+    def __instancecheck__(self, instance: object) -> bool:
+        if isinstance(instance, Decorator):
+            if instance.is_overload:
+                return issubclass(OverloadedFuncDef, self) and instance.is_callable
+            else:
+                return issubclass(FuncBase, self) and instance.is_callable
+        else:
+            return super().__instancecheck__(instance)
```

Comment:
Sorry, but this will be a big no from me. I don't like `Decorator` (which is an understatement), but this doesn't look like a solution:
* This will make future issues hard to debug.
* This will not work with `mypyc`.

There should be a better solution, which is probably just fixing several `isinstance()` checks for now to account for `Decorator` (also I am not sure there are many of them to be updated if we just want to fix `__init__`).

Reply:
Have implemented an alternate approach. You may like this one better.